lib/src/lib.rs

```diff
@@ -1134,194 +1092,111 @@ impl Repo {
-            match rx.recv().unwrap() {
-                ChunkAssemblerMessage::Exit => {
-                    tx.send(ChunkMessage::Exit).unwrap();
-                    assert!(previous_parts.is_empty());
-                    return;
-                }
-                ChunkAssemblerMessage::Data(part,
-                                            edges,
-                                            chunk_type,
-                                            data_type) => {
-                    if edges.is_empty() {
-                        previous_parts.push(part)
-                    } else {
-                        self.chunk_assembler_handle_data_with_edges(
-                            &mut tx,
-                            part,
-                            edges,
-                            chunk_type,
-                            data_type,
-                            &mut previous_parts)
-                    }
-                }
+        fn flush_all(queue: &mut VecDeque<(File, PathBuf)>,
+                     tx: &two_lock_queue::Sender<(File, PathBuf)>) {
+            while !queue.is_empty() {
+                flush_one(queue, tx);
             }
         }
-    }
 
-    fn chunk_compressor(&self,
-                        rx: mpsc::Receiver<ChunkMessage>,
-                        tx: mpsc::SyncSender<ChunkMessage>) {
-        loop {
-            match rx.recv().unwrap() {
-                ChunkMessage::Exit => {
-                    tx.send(ChunkMessage::Exit).unwrap();
-                    return;
-                }
-                ChunkMessage::Data(data, sha256, chunk_type, data_type) => {
-                    let tx_data = if data_type.should_compress() {
-                        let mut compressor =
-                            flate2::write::DeflateEncoder::new(
-                                Vec::with_capacity(data.len()),
-                                flate2::Compression::Default);
-
-                        compressor.write_all(&data).unwrap();
-                        compressor.finish().unwrap()
-                    } else {
-                        data
-                    };
-                    tx.send(ChunkMessage::Data(tx_data,
-                                                 sha256,
-                                                 chunk_type,
-                                                 data_type))
-                        .unwrap();
-                }
+        crossbeam::scope(|scope| {
+            // Unlike CPU-intense workload, it's hard to come up with one
+            // formula for best number of `fdatasync` threads for everyone.
+            // Each buffer is blocked on one chunk, so the longer the chunks,
+            // the less threads are needed to saturate the disk and so on. Some
+            // form of run-time scaling would be best. I picked some reasonable
+            // that also works well for me with defaults. --dpc
+            let thread_num = 4 * self.write_cpu_thread_num();
```

Comment:
A subject for future investigation. I 'd like to add logging-based performance reporting first so it's easier to figure out what works, and what not.